### DISCUSSION TITLE
_vt_HOLD_... tables and schema matching

### DISCUSSION THREAD

**Discussion by mhamann** [planetscale/discussion#517] [Q&A] [answered]
We've built a feature in our database pipeline that compares the current database schema with the expected schema before a change is allowed to be promoted to our staging or production environments. That way, we don't end up with a server deployment that doesn't match the database.

However, this is currently problematic, because sometimes the PlanetScale database has extra tables like `_vt_HOLD_0ada36c323aa51c5a56b59fc384cc8a6_20230815201031`. Of course, these aren't in the schema, so the diff check complains and fails our deployment.

First, why do these tables exist beyond the lifetime of a deployed branch? It seems they hold copies of data when a DR is modifying a table while allowing rollback. Once the rollback period expires, shouldn't these tables be cleaned up?

Second, is there a way to force PlanetScale to delete them or hide them from our view so that the DB migration script can't see them?

**Comment by mscoutermarsh** [comment, marked as answer]
Hi @mhamann,

Those are the "ghost" tables used for schema migrations. They get automatically cleaned up by Vitess after a few days. These docs give a good explainer on what they are doing: [How online schema change tools work](https://planetscale.com/docs/learn/how-online-schema-change-tools-work#initializing-the-ghost-table-schema). And for even more detail: [Vitess table lifecycle](https://vitess.io/docs/17.0/user-guides/schema-changes/table-lifecycle/).

Keeping them around for a couple days has two primary benefits: 1. If a mistake was made with a migration, the old copy of the table can still be accessed. 2. There can be performance impact on the database dropping a large table. 

I believe your best option here would be filtering them out of your script. Here's a regex you could use to match on the table names.

```ruby
VITESS_GHOST_TABLE_REGEX = "_vt_EVAC_.*|_vt_DROP_.*|_vt_PURGE_.*|_vt_HOLD_.*|_[0-9a-zA-Z]{8}_[0-9a-zA-Z]{4}_[0-9a-zA-Z]{4}_.*"
```
